src/webgpu/shader/validation/variable_and_const.spec.ts

```diff
@@ -0,0 +1,105 @@
+export const description = `
+Positive and negative validation tests for variable and const.
+`;
+
+import { params, poptions } from '../../../common/framework/params_builder.js';
+import { makeTestGroup } from '../../../common/framework/test_group.js';
+
+import { ShaderValidationTest } from './shader_validation_test.js';
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+const kScalarType = ['i32', 'f32', 'u32', 'bool'] as const;
+type ScalarType = 'i32' | 'f32' | 'u32' | 'bool';
+
+const kContainerTypes = [
+  undefined,
+  'vec2',
+  'vec3',
+  'vec4',
+  'mat2x2',
+  'mat2x3',
+  'mat2x4',
+  'mat3x2',
+  'mat3x3',
+  'mat3x4',
+  'mat4x2',
+  'mat4x3',
+  'mat4x4',
+  'array',
+] as const;
+type ContainerType =
+  | undefined
+  | 'vec2'
+  | 'vec3'
+  | 'vec4'
+  | 'mat2x2'
+  | 'mat2x3'
+  | 'mat2x4'
+  | 'mat3x2'
+  | 'mat3x3'
+  | 'mat3x4'
+  | 'mat4x2'
+  | 'mat4x3'
+  | 'mat4x4'
+  | 'array';
+
+function getType(scalarType: ScalarType, containerType: ContainerType) {
+  let type = '';
+  switch (containerType) {
+    case undefined: {
+      type = scalarType;
+      break;
+    }
+    case 'array': {
+      // TODO(sarahM0): 12 is a random number here. find a solution to replace it.
+      // TODO(sarahM0): test array of vectors and matrices. eg. array<vec4, 9>
+      type = `array<${scalarType}, 12>`;
+      break;
+    }
+    default: {
+      type = containerType ? `${containerType}<${scalarType}>` : scalarType;
+      break;
+    }
+  }
+  return type;
+}
+
+// TODO(sarahM0): v-0033: test structs
```

Comment:
(not a comment, just a discussion)

This came up in this week's call but I don't know the answer: are structs nominal or structural types? That is, is the following valid or invalid?:

```
struct A { x: i32; }
struct B { x: i32; }

const a: A = ...;
const b: B = a;
```

If it's valid (structs are structural), then we should extend the test here. If it's invalid (structs are nominal), it would make most sense to have a separate test for this that is just like what I wrote above.